src/main/java/edu/harvard/iq/dataverse/search/IndexServiceBean.java

```diff
@@ -348,48 +349,78 @@ public Future<String> indexDataverse(Dataverse dataverse, boolean processPaths)
     }
     
     @TransactionAttribute(REQUIRES_NEW)
-    public Future<String> indexDatasetInNewTransaction(Long datasetId) throws  SolrServerException, IOException{ //Dataset dataset) {
+    public void indexDatasetInNewTransaction(Long datasetId) { //Dataset dataset) {
         boolean doNormalSolrDocCleanUp = false;
         Dataset dataset = em.find(Dataset.class, datasetId);
-        // return indexDataset(dataset, doNormalSolrDocCleanUp);
-        Future<String> ret = indexDataset(dataset, doNormalSolrDocCleanUp);
+        asyncIndexDataset(dataset, doNormalSolrDocCleanUp);
         dataset = null;
-        return ret;
     }
+
+    // The following two variables are only used in the synchronized getNextToIndex method and do not need to be synchronized themselves
     
-    @TransactionAttribute(REQUIRES_NEW)
-    public Future<String> indexDatasetObjectInNewTransaction(Dataset dataset) throws  SolrServerException, IOException{ //Dataset dataset) {
-        boolean doNormalSolrDocCleanUp = false;
-        // return indexDataset(dataset, doNormalSolrDocCleanUp);
-        Future<String> ret = indexDataset(dataset, doNormalSolrDocCleanUp);
-        dataset = null;
-        return ret;
+    // nextToIndex contains datasets mapped by dataset id that were added for future indexing while the indexing was already ongoing for a given dataset
+    // (if there already was a dataset scheduled for indexing, it is overwritten and only the most recently requested version is kept in the map)
+    private static Map<Long, Dataset> nextToIndex = new HashMap<>();
+    // indexingNow is a set of dataset ids of datasets being indexed asynchronously right now
+    private static Set<Long> indexingNow = new HashSet<>();
+
+    // When you pass null as Dataset parameter to this method, it indicates that the indexing of the dataset with "id" has finished
+    // Pass non-null Dataset to schedule it for indexing
+    synchronized private static Dataset getNextToIndex(Long id, Dataset d) {
```

Comment:
What I do not like about this approach: it implements lots of custom logic to solve the producer-consumer problem at hand instead of using Java standard mechanisms from `java.util.concurrent`.

Reply:
I went with custom logic because of the specific scenario: I am doing many updates (e.g., uploading 1000s of files) one after another, and I want to index only once at the end. Since I can not know when the end would come, and the same solution must work for just one update where we want the indexing result to be available as soon as possible, I can only skip as many index operations as possible, but not more than that.

I explicitly did not want to use a queue and a producer-consumer logic, as it would make it very complex to achieve the same functionality as it is implemented now. I would still have to know if an index is being done for a specific dataset at a given time (indexingNow set) and I would need to implement removing items from the queue, and even worse: reordering of the queue to give a priority to the datasets not being indexed now. Synchronizing the producer-consumer logic the way the current solution does would also be very difficult.

Reply:
The existing queue mechanism provide all this...?

> I would still have to know if an index is being done for a specific dataset at a given time (indexingNow set)

Why would this be necessary? The executors would take from the queue (so when a dataset is being indexed right now it not be in the queue at the time you want to edit the same dataset again). The problem to solve is to make another executor not take up the same dataset as long as it is being indexed by another. There are 2 options for that: a) set a `Lock` or a `Semaphore` and/or b) add a delay for the taking (readily avail as `DelayedQueue`) where delay time > index time

>  and I would need to implement removing items from the queue, 

Nope, you will only need to check if a dataset is in the queue already and not add it again.

> and even worse: reordering of the queue to give a priority to the datasets not being indexed now.

I don't think that is necessary.

> Synchronizing the producer-consumer logic the way the current solution does would also be very difficult.

Synchronizing what exactly? Are you referring to the `synchronized` modifier? The blocking queues in `java.util.concurrent` are thread safe already.

Reply:
DelayQueue would not work, as index of empty dataset takes a second, and index of a dataset with 7000 files takes 5 minutes (we don't know yet how long it takes for 50000 files as it is not possible with the current code to upload that many in less than 270 days).

The Lock you mention is exactly what I said about the synchronizing, regardless of the fact that the queue is already synchronized. To be efficient and allow indexing of multiple datasets at once, you would need a lock for each dataset. To have a constant time of finding locks, you could store them in a set (exactly what I am doing now with the set indexingNow).

Checking if a dataset is already in a queue is done now with contains, which is a constant time operation for a map. The new solution would look almost the same as the current one, except for the expensive (i.e., not constant time) queue operations of contains and reordering, which would be very ugly and hard to implement.

And reordering makes sense even if there is only one dataset in the queue (or even stronger: especially when there is only one dataset in the queue). If it is the one that takes 5 minutes, some other user would need to wait at least 5 minutes before his dataset would be indexed.

This problem gets worse when we have multiple workers: either they pick a dataset from a queue and wait for the lock to get free (which means that adding a file to that dataset would put another indexing for that dataset in the queue, etc., until all workers wait on the same lock to index the same dataset), or they wait until locked dataset goes away from the queue before picking the next one (leaving the workers with nothing to do for 5 minutes).

A way better solution would be then that workers pick up only datasets that are not locked now. At that point, a queue does not make much sense and can be better replaced with a map having constant operation times. Exactly what happens now.

I have really considered a queue before I implemented this and have given a lot of thought to that problem. It might not look like much, but I believe that my code is a nice and elegant solution to a very hard problem. Please reconsider it as it is now. I am used to using queues in my code (you can even check my integration tool, I think it handles concurrency very nicely with a queue), I just do not see how it fits here. I might be missing something, and will look at it again, but I think I am already biased...

Reply:
Thank you @ErykKul for this explanation! Glad you already gave this that much thought, much appreciated!

I will stop arguing here, as you make some very valid points with the non-linearity of index times.

May I request that you do add explanations about this to the code? All of this seems to be non-trivial and we have no good other place to document architectural decisions aside from the code itself. Some in-depth Javadoc would certainly help to clarify what has been tried and why this is the solution we go for. In case we revisit this in the future, this would provide very valuable insights that might be forgotten about by then.

Thank you for working on this! Much appreciated!

Reply:
Great, thanks! I also appreciate all your comments and input, I have still a lot to learn about the EJBs, the existing code, etc. I will extend the documentation to clarify the intent (I can see why this code might rise some eyebrows at the first sight). Maybe we will figure out something better in the long term.